tools/feature_matrix.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2022 The Emscripten Authors.  All rights reserved.
+# Emscripten is available under two separate licenses, the MIT license and the
+# University of Illinois/NCSA Open Source License.  Both these licenses can be
+# found in the LICENSE file.
+
+"""Utilities for mapping browser versions to llvm feature flags."""
+
+import logging
+from enum import IntEnum, auto
+
+from .settings import settings
+
+logger = logging.getLogger('feature_matrix')
+
+
+class Feature(IntEnum):
+  NON_TRAPPING_FPTOINT = auto()
+  SIGN_EXT = auto()
+  BULK_MEMORY = auto()
+  MUTABLE_GLOBALS = auto()
+
+
+min_browser_versions = {
+  Feature.NON_TRAPPING_FPTOINT: {
+    'chrome': 75,
+    'firefox': 65,
+    'safari': 150000,
+  },
+  Feature.SIGN_EXT: {
+    'chrome': 74,
+    'firefox': 62,
+    'safari': 141000,
+  },
+  Feature.BULK_MEMORY: {
+    'chrome': 75,
+    'firefox': 79,
+    'safari': 150000,
+  },
+  Feature.MUTABLE_GLOBALS: {
+    'chrome': 74,
+    'firefox': 61,
+    'safari': 120000,
+  },
+}
+
+
+def caniuse(feature):
+  min_versions = min_browser_versions[feature]
+  if settings.MIN_CHROME_VERSION < min_versions['chrome']:
+    return False
+  if settings.MIN_FIREFOX_VERSION < min_versions['firefox']:
+    return False
+  if settings.MIN_SAFARI_VERSION < min_versions['safari']:
+    return False
+  # IE and Edge don't support any non-MVP features
+  if settings.MIN_IE_VERSION != 0x7FFFFFFF or settings.MIN_EDGE_VERSION != 0x7FFFFFFF:
+    return False
+  return True
+
+
+def get_feature_flags():
+  flags = []
+
+  # Enabling threads, or PIC implicitly opts into many of these features.
+  # TODO(sbc): We should probably have USE_PTHREADS/RELOCATABLE
+  # implicitly bump the min browser versions instead.
```

Comment:
I think this TODO would be a good idea. Here is where we do it already for bigint:

https://github.com/emscripten-core/emscripten/blob/0fe8aa8e91a77aea12068e31ff4c5d3ee44a497c/emcc.py#L401-L413

Reply:
This is somewhat surprising behavior (not the linking of pthreads with dependent features, but the upgrading of all features because one of them requires a particular browser version). It sort of makes sense if the user is targeting a known browser, but there could be use cases where you'd want to let the user override emscripten's choices. Most of them are variations of "the user knows more about the feature matrix than emscripten does." for example:  using a browser (or other JS runtime) that is unknown to emscripten, using an environment that has features removed for security, or even added (something like electron?). All of these might have different feature sets so the user would probably either pick a browser baseline and add features to it, or just manually select features. (I guess it also conflicts with the behavior that adding clang flags like `-mmultivalue` does not affect emscripten's choice of target versions)